In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import os
from collections import Counter
# print(os.listdir('data.csv'))

In [2]:
data = pd.read_csv('data.csv')
data.head(5)

imdb_id  popularity     budget     revenue                original_title  \
0  tt0369610   32.985763  150000000  1513528810                Jurassic World   
1  tt1392190   28.419936  150000000   378436354            Mad Max: Fury Road   
2  tt2908446   13.112507  110000000   295238201                     Insurgent   
3  tt2488496   11.173104  200000000  2068178225  Star Wars: The Force Awakens   
4  tt2820852    9.335014  190000000  1506249360                     Furious 7   

                                                cast          director  \
0  Chris Pratt|Bryce Dallas Howard|Irrfan Khan|Vi...   Colin Trevorrow   
1  Tom Hardy|Charlize Theron|Hugh Keays-Byrne|Nic...     George Miller   
2  Shailene Woodley|Theo James|Kate Winslet|Ansel...  Robert Schwentke   
3  Harrison Ford|Mark Hamill|Carrie Fisher|Adam D...       J.J. Abrams   
4  Vin Diesel|Paul Walker|Jason Statham|Michelle ...         James Wan   

                         tagline  \
0              The park is open.   
1             What a Lovely Day.   
2     One Choice Can Destroy You   
3  Every generation has a story.   
4            Vengeance Hits Home   

                                            overview  runtime  \
0  Twenty-two years after the events of Jurassic ...      124   
1  An apocalyptic story set in the furthest reach...      120   
2  Beatrice Prior must confront her inner demons ...      119   
3  Thirty years after defeating the Galactic Empi...      136   
4  Deckard Shaw seeks revenge against Dominic Tor...      137   

                                      genres  \
0  Action|Adventure|Science Fiction|Thriller   
1  Action|Adventure|Science Fiction|Thriller   
2         Adventure|Science Fiction|Thriller   
3   Action|Adventure|Science Fiction|Fantasy   
4                      Action|Crime|Thriller   

                                production_companies release_date  vote_count  \
0  Universal Studios|Amblin Entertainment|Legenda...     6/9/2015        5562   
1  Village Roadshow Pictures|Kennedy Miller Produ...    5/13/2015        6185   
2  Summit Entertainment|Mandeville Films|Red Wago...    3/18/2015        2480   
3          Lucasfilm|Truenorth Productions|Bad Robot   12/15/2015        5292   
4  Universal Pictures|Original Film|Media Rights ...     4/1/2015        2947   

   vote_average  release_year  
0           6.5          2015  
1           7.1          2015  
2           6.3          2015  
3           7.5          2015  
4           7.3          2015

In [3]:
len(data)

1890

# Предобработка датасета

In [4]:
answer_ls = [] # создадим список с ответами. сюда будем добавлять ответы по мере прохождения теста
# сюда можем вписать создание новых колонок в датасете

data['profit'] = data['revenue'] - data['budget']
data['release_month'] = data.release_date.apply(lambda x:x.split('/')[0])
data['len_title'] = data.original_title.apply(lambda x: len(x))
data['len_word'] = data.original_title.apply(lambda x: len(x.split(' ')))

# 1. У какого фильма из списка самый большой бюджет?
Варианты ответов:
1. The Dark Knight Rises (tt1345836)
2. Spider-Man 3 (tt0413300)
3. Avengers: Age of Ultron (tt2395427)
4. The Warrior's Way	(tt1032751)
5. Pirates of the Caribbean: On Stranger Tides (tt1298650)

In [5]:
data[data.budget == data.budget.max()].original_title

491    The Warrior's Way
Name: original_title, dtype: object

In [6]:
data.groupby(['original_title'])['budget'].max().sort_values(ascending=False).head()

original_title
The Warrior's Way                              425000000
Pirates of the Caribbean: On Stranger Tides    380000000
Pirates of the Caribbean: At World's End       300000000
Avengers: Age of Ultron                        280000000
John Carter                                    260000000
Name: budget, dtype: int64

In [7]:
# тут вводим ваш ответ и добавлем в его список ответов (сейчас для примера стоит "1")
answer_ls.append(4)

# 2. Какой из фильмов самый длительный (в минутах)
1. The Lord of the Rings: The Return of the King	(tt0167260)
2. Gods and Generals	(tt0279111)
3. King Kong	(tt0360717)
4. Pearl Harbor	(tt0213149)
5. Alexander	(tt0346491)

In [8]:
data[data.runtime == data.runtime.max()]

imdb_id  popularity    budget   revenue     original_title  \
1158  tt0279111    0.469518  56000000  12923936  Gods and Generals   

                                                   cast           director  \
1158  Stephen Lang|Jeff Daniels|Robert Duvall|Kevin ...  Ronald F. Maxwell   

                                  tagline  \
1158  The nations heart was touched by...   

                                               overview  runtime  \
1158  The film centers mostly around the personal an...      214   

                 genres                production_companies release_date  \
1158  Drama|History|War  Turner Pictures|Antietam Filmworks    2/21/2003   

      vote_count  vote_average  release_year    profit release_month  \
1158          23           5.8          2003 -43076064             2   

      len_title  len_word  
1158         17         3

In [9]:
data.groupby(['original_title'])['runtime'].max().sort_values(ascending=False).head()

original_title
Gods and Generals                                214
The Lord of the Rings: The Return of the King    201
Grindhouse                                       191
King Kong                                        187
Pearl Harbor                                     183
Name: runtime, dtype: int64

In [10]:
answer_ls.append(2)

# 3. Какой из фильмов самый короткий (в минутах)
Варианты ответов:

1. Home on the Range	tt0299172
2. The Jungle Book 2	tt0283426
3. Winnie the Pooh	tt1449283
4. Corpse Bride	tt0121164
5. Hoodwinked!	tt0443536

In [11]:
data.sort_values(by = 'runtime').head(1)

imdb_id  popularity    budget   revenue   original_title  \
769  tt1449283    1.425344  30000000  14460000  Winnie the Pooh   

                                                  cast  \
769  Jim Cummings|Travis Oates|Jim Cummings|Bud Luc...   

                      director   tagline  \
769  Stephen Anderson|Don Hall  Oh Pooh.   

                                              overview  runtime  \
769  During an ordinary day in Hundred Acre Wood, W...       63   

               genres                               production_companies  \
769  Animation|Family  Walt Disney Pictures|Walt Disney Animation Stu...   

    release_date  vote_count  vote_average  release_year    profit  \
769    4/13/2011         174           6.8          2011 -15540000   

    release_month  len_title  len_word  
769             4         15         3

In [12]:
data.groupby(['original_title'])['runtime'].max().sort_values().head()

original_title
Winnie the Pooh         63
The Jungle Book 2       72
Return to Never Land    72
Home on the Range       76
Valiant                 76
Name: runtime, dtype: int64

In [13]:
data[data.runtime == data.runtime.min()]

imdb_id  popularity    budget   revenue   original_title  \
769  tt1449283    1.425344  30000000  14460000  Winnie the Pooh   

                                                  cast  \
769  Jim Cummings|Travis Oates|Jim Cummings|Bud Luc...   

                      director   tagline  \
769  Stephen Anderson|Don Hall  Oh Pooh.   

                                              overview  runtime  \
769  During an ordinary day in Hundred Acre Wood, W...       63   

               genres                               production_companies  \
769  Animation|Family  Walt Disney Pictures|Walt Disney Animation Stu...   

    release_date  vote_count  vote_average  release_year    profit  \
769    4/13/2011         174           6.8          2011 -15540000   

    release_month  len_title  len_word  
769             4         15         3

In [14]:
answer_ls.append(3)

# 4. Какое число ближе к средней длительности фильма в датасете?

Варианты ответов:
1. 115
2. 110
3. 105
4. 120
5. 100


In [15]:
data.runtime.mean()

109.65343915343915

In [16]:
answer_ls.append(2)

# 5. Какое число ближе к медианной длительности фильма в датасете?
Варианты ответов:
1. 106
2. 112
3. 101
4. 120
5. 115




In [17]:
data.runtime.median()

106.5

In [18]:
answer_ls.append(1)

# 6. Какой самый прибыльный фильм?
Варианты ответов:
1. The Avengers	tt0848228
2. Minions	tt2293640
3. Star Wars: The Force Awakens	tt2488496
4. Furious 7	tt2820852
5. Avatar	tt0499549

In [19]:
data[data.profit == data.profit.max()]

imdb_id  popularity     budget     revenue original_title  \
239  tt0499549    9.432768  237000000  2781505847         Avatar   

                                                  cast       director  \
239  Sam Worthington|Zoe Saldana|Sigourney Weaver|S...  James Cameron   

                         tagline  \
239  Enter the World of Pandora.   

                                              overview  runtime  \
239  In the 22nd century, a paraplegic Marine is di...      162   

                                       genres  \
239  Action|Adventure|Fantasy|Science Fiction   

                                  production_companies release_date  \
239  Ingenious Film Partners|Twentieth Century Fox ...   12/10/2009   

     vote_count  vote_average  release_year      profit release_month  \
239        8458           7.1          2009  2544505847            12   

     len_title  len_word  
239          6         1

In [20]:
answer_ls.append(5)

# 7. Какой фильм самый убыточный?
Варианты ответов:
1. Supernova tt0134983
2. The Warrior's Way tt1032751
3. Flushed Away	tt0424095
4. The Adventures of Pluto Nash	tt0180052
5. The Lone Ranger	tt1210819

In [21]:
data[data.profit == data.profit.min()]

imdb_id  popularity     budget   revenue     original_title  \
491  tt1032751     0.25054  425000000  11087569  The Warrior's Way   

                                                  cast    director  \
491  Kate Bosworth|Jang Dong-gun|Geoffrey Rush|Dann...  Sngmoo Lee   

                     tagline  \
491  Assassin. Hero. Legend.   

                                              overview  runtime  \
491  An Asian assassin (Dong-gun Jang) is forced to...      100   

                                        genres      production_companies  \
491  Adventure|Fantasy|Action|Western|Thriller  Boram Entertainment Inc.   

    release_date  vote_count  vote_average  release_year     profit  \
491    12/2/2010          74           6.4          2010 -413912431   

    release_month  len_title  len_word  
491            12         17         3

In [22]:
answer_ls.append(2)

# 8. У скольких фильмов из датасета объем сборов оказался выше бюджета?
Варианты ответов:
1. 1478
2. 1520
3. 1241
4. 1135
5. 1398


In [23]:
len(data[data.profit > 0])

1478

In [24]:
answer_ls.append(1)

# 9. Самый прибыльный фильм в 2008 году?
Варианты ответов:
1. Madagascar: Escape 2 Africa	tt0479952
2. Iron Man	tt0371746
3. Kung Fu Panda	tt0441773
4. The Dark Knight	tt0468569
5. Mamma Mia!	tt0795421

In [25]:
data_2008=data[data.release_year == 2008]
data_2008.sort_values(by = ['profit'], ascending=False).head(1).original_title

600    The Dark Knight
Name: original_title, dtype: object

In [26]:
answer_ls.append(4)

# 10. Самый убыточный фильм за период с 2012 по 2014 (включительно)?
Варианты ответов:
1. Winter's Tale	tt1837709
2. Stolen	tt1656186
3. Broken City	tt1235522
4. Upside Down	tt1374992
5. The Lone Ranger	tt1210819


In [27]:
data10 = data[(data.release_year >= 2012) & (data.release_year <= 2014)]
data10.sort_values(by = ['profit']).original_title.head(1)

1246    The Lone Ranger
Name: original_title, dtype: object

In [28]:
answer_ls.append(5)

# 11. Какого жанра фильмов больше всего?
Варианты ответов:
1. Action
2. Adventure
3. Drama
4. Comedy
5. Thriller

In [29]:
data11= data.genres
Counter(data11.str.cat(sep='|').split('|'))

Counter({'Action': 583,
         'Adventure': 416,
         'Science Fiction': 248,
         'Thriller': 597,
         'Fantasy': 223,
         'Crime': 315,
         'Western': 20,
         'Drama': 782,
         'Family': 260,
         'Animation': 139,
         'Comedy': 683,
         'Mystery': 168,
         'Romance': 308,
         'War': 58,
         'History': 62,
         'Music': 64,
         'Horror': 176,
         'Documentary': 8,
         'Foreign': 2})

In [30]:
pd.Series(data['genres'].str.cat(sep='|').split('|')).value_counts()

Drama              782
Comedy             683
Thriller           597
Action             583
Adventure          416
Crime              315
Romance            308
Family             260
Science Fiction    248
Fantasy            223
Horror             176
Mystery            168
Animation          139
Music               64
History             62
War                 58
Western             20
Documentary          8
Foreign              2
dtype: int64

In [31]:
answer_ls.append(3)

# 12. Какого жанра среди прибыльных фильмов больше всего?
Варианты ответов:
1. Drama
2. Comedy
3. Action
4. Thriller
5. Adventure

In [32]:
data12 = data[data.profit > 0]
data12 = data12.genres
Counter(data12.str.cat(sep='|').split('|'))


Counter({'Action': 444,
         'Adventure': 337,
         'Science Fiction': 195,
         'Thriller': 446,
         'Fantasy': 188,
         'Crime': 231,
         'Western': 12,
         'Drama': 560,
         'Family': 226,
         'Animation': 120,
         'Comedy': 551,
         'Mystery': 119,
         'Romance': 242,
         'War': 41,
         'History': 46,
         'Music': 47,
         'Horror': 150,
         'Documentary': 7})

In [33]:
data12 = data[data.profit > 0]
pd.Series(data12['genres'].str.cat(sep='|').split('|')).value_counts()

Drama              560
Comedy             551
Thriller           446
Action             444
Adventure          337
Romance            242
Crime              231
Family             226
Science Fiction    195
Fantasy            188
Horror             150
Animation          120
Mystery            119
Music               47
History             46
War                 41
Western             12
Documentary          7
dtype: int64

In [34]:
answer_ls.append(1)

# 13. Кто из режиссеров снял больше всего фильмов?
Варианты ответов:
1. Steven Spielberg
2. Ridley Scott 
3. Steven Soderbergh
4. Christopher Nolan
5. Clint Eastwood

In [35]:
pd.Series(data['director'].str.cat(sep='|').split('|')).value_counts().head()

Steven Soderbergh    13
Clint Eastwood       12
Ridley Scott         12
Robert Rodriguez     11
Steven Spielberg     10
dtype: int64

In [36]:
answer_ls.append(3)

# 14. Кто из режиссеров снял больше всего Прибыльных фильмов?
Варианты ответов:
1. Steven Soderbergh
2. Clint Eastwood
3. Steven Spielberg
4. Ridley Scott
5. Christopher Nolan

In [37]:
data14 = data[data.profit > 0]
pd.Series(data14['director'].str.cat(sep='|').split('|')).value_counts().head()

Ridley Scott         12
Steven Soderbergh    10
Clint Eastwood       10
Steven Spielberg     10
Tim Burton            9
dtype: int64

In [38]:
answer_ls.append(4)

# 15. Кто из режиссеров принес больше всего прибыли?
Варианты ответов:
1. Steven Spielberg
2. Christopher Nolan
3. David Yates
4. James Cameron
5. Peter Jackson


In [39]:
data15 = data[data.profit > 0][['director', 'profit']]
data15 = pd.DataFrame(data15['director'].str.split('|').tolist(), index = data15['profit']).stack().reset_index([0])
data15.columns = ['profit', 'director']
data15.groupby(['director'])['profit'].sum().sort_values(ascending=False).head()

director
Peter Jackson        5202593685
David Yates          3379295625
Christopher Nolan    3162548502
J.J. Abrams          2839169916
Michael Bay          2760938960
Name: profit, dtype: int64

In [40]:
answer_ls.append(5)

# 16. Какой актер принес больше всего прибыли?
Варианты ответов:
1. Emma Watson
2. Johnny Depp
3. Michelle Rodriguez
4. Orlando Bloom
5. Rupert Grint

In [41]:
data16 = data[data.profit > 0][['cast', 'profit']]
data16 = pd.DataFrame(data16['cast'].str.split('|').tolist(), index = data16['profit']).stack().reset_index([0])
data16.columns = ['profit', 'actor']
data16.groupby(['actor'])['profit'].sum().sort_values(ascending=False).head()

actor
Emma Watson          6675368452
Daniel Radcliffe     6523915983
Rupert Grint         6413185247
Ian McKellen         6171916461
Robert Downey Jr.    5361847918
Name: profit, dtype: int64

In [42]:
answer_ls.append(1)

# 17. Какой актер принес меньше всего прибыли в 2012 году?
Варианты ответов:
1. Nicolas Cage
2. Danny Huston
3. Kirsten Dunst
4. Jim Sturgess
5. Sami Gayle

In [43]:
data17 = data[(data.profit < 0) & (data.release_year == 2012)][['cast', 'profit']]
data17 = pd.DataFrame(data17['cast'].str.split('|').tolist(), index = data17['profit']).stack().reset_index([0])
data17.columns = ['profit', 'actor']
data17.groupby(['actor'])['profit'].sum().sort_values().head()

actor
Kirsten Dunst   -68109207
James Kidnie    -51893525
Heidi Hawkins   -51893525
Jim Sturgess    -51893525
Timothy Spall   -51893525
Name: profit, dtype: int64

In [44]:
answer_ls.append(3)

# 18. Какой актер снялся в большем количестве высокобюджетных фильмов? (в фильмах где бюджет выше среднего по данной выборке)
Варианты ответов:
1. Tom Cruise
2. Mark Wahlberg 
3. Matt Damon
4. Angelina Jolie
5. Adam Sandler

In [45]:
data18 = data[data.budget > data.budget.mean()]
pd.Series(data18['cast'].str.cat(sep='|').split('|')).value_counts().head()

Matt Damon           18
Adam Sandler         17
Angelina Jolie       16
Samuel L. Jackson    15
Tom Cruise           15
dtype: int64

In [46]:
answer_ls.append(3)

# 19. В фильмах какого жанра больше всего снимался Nicolas Cage?  
Варианты ответа:
1. Drama
2. Action
3. Thriller
4. Adventure
5. Crime

In [47]:
data19 = data[data['cast'].str.contains('Nicolas Cage')]
pd.Series(data19['genres'].str.cat(sep='|').split('|')).value_counts().head()

Action      17
Thriller    15
Drama       12
Crime       10
Fantasy      8
dtype: int64

In [48]:
answer_ls.append(2)

# 20. Какая студия сняла больше всего фильмов?
Варианты ответа:
1. Universal Pictures (Universal)
2. Paramount Pictures
3. Columbia Pictures
4. Warner Bros
5. Twentieth Century Fox Film Corporation

In [49]:
pd.Series(data['production_companies'].str.cat(sep='|').split('|')).value_counts().head()

Universal Pictures                        173
Warner Bros.                              168
Paramount Pictures                        122
Columbia Pictures                         117
Twentieth Century Fox Film Corporation    109
dtype: int64

In [50]:
answer_ls.append(1)

# 21. Какая студия сняла больше всего фильмов в 2015 году?
Варианты ответа:
1. Universal Pictures
2. Paramount Pictures
3. Columbia Pictures
4. Warner Bros
5. Twentieth Century Fox Film Corporation

In [51]:
data21 = data[data.release_year == 2015]
pd.Series(data21['production_companies'].str.cat(sep='|').split('|')).value_counts().head()

Warner Bros.                              12
Universal Pictures                        10
Twentieth Century Fox Film Corporation     8
Columbia Pictures                          7
Paramount Pictures                         7
dtype: int64

In [52]:
answer_ls.append(4)

# 22. Какая студия заработала больше всего денег в жанре комедий за все время?
Варианты ответа:
1. Warner Bros
2. Universal Pictures (Universal)
3. Columbia Pictures
4. Paramount Pictures
5. Walt Disney

In [53]:
data22 = data[data['genres'].str.contains('Comedy')]
data22 = data22.loc[:, ['production_companies', 'profit']]
data22 = pd.DataFrame(data22['production_companies'].str.split('|').tolist(), index = data22['profit']).stack().reset_index([0])
data22.columns = ['profit', 'comp']
data22.groupby(['comp'])['profit'].sum().sort_values(ascending=False).head()

comp
Universal Pictures                        8961545581
Walt Disney Pictures                      7669710326
Twentieth Century Fox Film Corporation    5686960294
Columbia Pictures                         5646343696
DreamWorks Animation                      4789049764
Name: profit, dtype: int64

In [54]:
answer_ls.append(2)

# 23. Какая студия заработала больше всего денег в 2012 году?
Варианты ответа:
1. Universal Pictures (Universal)
2. Warner Bros
3. Columbia Pictures
4. Paramount Pictures
5. Lucasfilm

In [55]:
data23 = data[data.release_year == 2012][['production_companies', 'profit']]
data23 = pd.DataFrame(data23['production_companies'].str.split('|').tolist(), index = data23['profit']).stack().reset_index([0])
data23.columns = ['profit', 'production_companies']
data23.groupby(['production_companies'])['profit'].sum().sort_values(ascending=False).head()

production_companies
Columbia Pictures     2501406608
Universal Pictures    1981011579
Marvel Studios        1299557910
Warner Bros.          1258020056
Relativity Media      1032593938
Name: profit, dtype: int64

In [56]:
answer_ls.append(3)

# 24. Самый убыточный фильм от Paramount Pictures
Варианты ответа:

1. K-19: The Widowmaker tt0267626
2. Next tt0435705
3. Twisted tt0315297
4. The Love Guru tt0811138
5. The Fighter tt0964517

In [57]:
data24 = data[data['production_companies'].str.contains('Paramount Pictures')]
data24.loc[data24['profit'] == data24.profit.min()].original_title

926    K-19: The Widowmaker
Name: original_title, dtype: object

In [58]:
answer_ls.append(1)

# 25. Какой Самый прибыльный год (заработали больше всего)?
Варианты ответа:
1. 2014
2. 2008
3. 2012
4. 2002
5. 2015

In [59]:
data.groupby(['release_year'])['profit'].sum().sort_values(ascending=False).head()

release_year
2015    18668572378
2014    16397812953
2012    16077001687
2013    15243179791
2011    14730241341
Name: profit, dtype: int64

In [60]:
answer_ls.append(5)

# 26. Какой Самый прибыльный год для студии Warner Bros?
Варианты ответа:
1. 2014
2. 2008
3. 2012
4. 2010
5. 2015

In [61]:
data26 = data[data['production_companies'].str.contains('Warner Bros')]
data26.groupby(['release_year'])['profit'].sum().sort_values(ascending=False).head()

release_year
2014    2295464519
2007    2201675217
2008    2134595031
2010    1974712985
2011    1871393682
Name: profit, dtype: int64

In [62]:
answer_ls.append(1)

# 27. В каком месяце за все годы суммарно вышло больше всего фильмов?
Варианты ответа:
1. Январь
2. Июнь
3. Декабрь
4. Сентябрь
5. Май

In [63]:
data.groupby(['release_month'])['imdb_id'].count().sort_values(ascending=False).head()

release_month
9     227
12    191
10    186
8     161
3     156
Name: imdb_id, dtype: int64

In [64]:
answer_ls.append(4)

# 28. Сколько суммарно вышло фильмов летом? (за июнь, июль, август)
Варианты ответа:
1. 345
2. 450
3. 478
4. 523
5. 381

In [65]:
# На этом моменте я вспомнил про регулярные выражения и решил попробовать их использовать.
len(data.query('release_month in ["6", "7", "8"]'))

450

In [66]:
answer_ls.append(2)

# 29. Какой режисер выпускает (суммарно по годам) больше всего фильмов зимой?
Варианты ответов:
1. Steven Soderbergh
2. Christopher Nolan
3. Clint Eastwood
4. Ridley Scott
5. Peter Jackson

In [67]:
data29 = data.query('release_month in ["1", "2", "12"]')
pd.Series(data29['director'].str.cat(sep='|').split('|')).value_counts().head()

Peter Jackson        7
Steven Soderbergh    6
Clint Eastwood       6
Nancy Meyers         4
Adam Shankman        4
dtype: int64

In [68]:
answer_ls.append(5)

# 30. Какой месяц чаще всего по годам самый прибыльный?
Варианты ответа:
1. Январь
2. Июнь
3. Декабрь
4. Сентябрь
5. Май

In [69]:
data.groupby(['release_month'])['profit'].sum().sort_values(ascending=False)

release_month
6     27551262500
12    26563071474
5     25039516141
11    18654155473
7     18307895020
3     13720216540
10    12994907859
4     12876005537
9     10878921041
8      9640610446
2      8854574425
1      5396885454
Name: profit, dtype: int64

In [70]:
answer_ls.append(2)

# 31. Названия фильмов какой студии в среднем самые длинные по количеству символов?
Варианты ответа:
1. Universal Pictures (Universal)
2. Warner Bros
3. Jim Henson Company, The
4. Paramount Pictures
5. Four By Two Productions

In [71]:
data31 = data[['len_title', 'production_companies']]
data31 = pd.DataFrame(data31['production_companies'].str.split('|').tolist(), index = data31['len_title']).stack().reset_index([0])
data31.columns = ['len_title', 'companies']
data31.groupby(['companies'])['len_title'].mean().sort_values(ascending=False).head()

companies
Four By Two Productions      83.0
Jim Henson Company, The      59.0
Dos Corazones                47.0
Polsky Films                 46.0
Museum Canada Productions    46.0
Name: len_title, dtype: float64

In [72]:
answer_ls.append(5)

# 32. Названия фильмов какой студии в среднем самые длинные по количеству слов?
Варианты ответа:
1. Universal Pictures (Universal)
2. Warner Bros
3. Jim Henson Company, The
4. Paramount Pictures
5. Four By Two Productions

In [73]:
data32 = data[['len_word', 'production_companies']]
data32 = pd.DataFrame(data32['production_companies'].str.split('|').tolist(), index = data32['len_word']).stack().reset_index([0])
data32.columns = ['len_word', 'companies']
data32.groupby(['companies'])['len_word'].mean().sort_values(ascending=False).head()

companies
Four By Two Productions    12.0
Jim Henson Company, The    10.0
Polsky Films                9.0
The Saul Zaentz Company     9.0
Dos Corazones               9.0
Name: len_word, dtype: float64

In [74]:
answer_ls.append(5)

# 33. Сколько разных слов используется в названиях фильмов?(без учета регистра)
Варианты ответа:
1. 6540
2. 1002
3. 2461
4. 28304
5. 3432

In [75]:
words=pd.Series(data['original_title'].str.cat(sep=' ').lower().split())
len(words.unique())

2461

In [76]:
answer_ls.append(3)

# 34. Какие фильмы входят в 1 процент лучших по рейтингу?
Варианты ответа:
1. Inside Out, Gone Girl, 12 Years a Slave
2. BloodRayne, The Adventures of Rocky & Bullwinkle
3. The Lord of the Rings: The Return of the King
4. 300, Lucky Number Slevin

In [77]:
data[(data.vote_average > np.percentile(data['vote_average'], 99))]

imdb_id  popularity     budget     revenue  \
9     tt2096673    6.326804  175000000   853708609   
34    tt3170832    3.557846    6000000    35401758   
118   tt0816692   24.949134  165000000   621752480   
119   tt2015381   14.311205  170000000   773312399   
125   tt2084970    8.110711   14000000   233555708   
128   tt2267998    6.438727   61000000   369330363   
138   tt2278388    4.930820   30000000   174600318   
370   tt1375666    9.363643  160000000   825500000   
600   tt0468569    8.466668  185000000  1001921825   
873   tt0253474    2.364204   35000000   120072577   
1082  tt0167260    7.122455   94000000  1118888979   
1184  tt0993846    4.877927  100000000   392000694   
1192  tt2024544    3.625529   20000000   187000000   
1801  tt0209144    3.095625    9000000    39723096   

                                     original_title  \
9                                        Inside Out   
34                                             Room   
118                                    Interstellar   
119                         Guardians of the Galaxy   
125                              The Imitation Game   
128                                       Gone Girl   
138                        The Grand Budapest Hotel   
370                                       Inception   
600                                 The Dark Knight   
873                                     The Pianist   
1082  The Lord of the Rings: The Return of the King   
1184                        The Wolf of Wall Street   
1192                               12 Years a Slave   
1801                                        Memento   

                                                   cast           director  \
9     Amy Poehler|Phyllis Smith|Richard Kind|Bill Ha...        Pete Docter   
34    Brie Larson|Jacob Tremblay|Joan Allen|Sean Bri...   Lenny Abrahamson   
118   Matthew McConaughey|Jessica Chastain|Anne Hath...  Christopher Nolan   
119   Chris Pratt|Zoe Saldana|Dave Bautista|Vin Dies...         James Gunn   
125   Benedict Cumberbatch|Keira Knightley|Matthew G...      Morten Tyldum   
128   Ben Affleck|Rosamund Pike|Carrie Coon|Neil Pat...      David Fincher   
138   Ralph Fiennes|Tony Revolori|F. Murray Abraham|...       Wes Anderson   
370   Leonardo DiCaprio|Joseph Gordon-Levitt|Ellen P...  Christopher Nolan   
600   Christian Bale|Michael Caine|Heath Ledger|Aaro...  Christopher Nolan   
873   Adrien Brody|Thomas Kretschmann|Frank Finlay|M...     Roman Polanski   
1082  Elijah Wood|Ian McKellen|Viggo Mortensen|Liv T...      Peter Jackson   
1184  Leonardo DiCaprio|Jonah Hill|Margot Robbie|Kyl...    Martin Scorsese   
1192  Chiwetel Ejiofor|Michael Fassbender|Lupita Nyo...      Steve McQueen   
1801  Guy Pearce|Carrie-Anne Moss|Joe Pantoliano|Mar...  Christopher Nolan   

                                                tagline  \
9              Meet the little voices inside your head.   
34                             Love knows no boundaries   
118   Mankind was born on Earth. It was never meant ...   
119                         All heroes start somewhere.   
125   The true enigma was the man who cracked the code.   
128         You don't know what you've got 'til it's...   
138             A perfect holiday without leaving home.   
370                Your mind is the scene of the crime.   
600                                     Why So Serious?   
873   Music was his passion. Survival was his master...   
1082                    The eye of the enemy is moving.   
1184                                EARN. SPEND. PARTY.   
1192    The extraordinary true story of Solomon Northup   
1801                  Some memories are best forgotten.   

                                               overview  runtime  \
9     Growing up can be a bumpy road, and it's no ex...       94   
34    Jack is a young boy of 5 years old who has liv...      117   
118   Interstellar chronicles the adventures of a gr...      169   
119   Light years from Earth, 26 years aft

In [78]:
answer_ls.append(1)

# 35. Какие актеры чаще всего снимаются в одном фильме вместе
Варианты ответа:
1. Johnny Depp & Helena Bonham Carter
2. Hugh Jackman & Ian McKellen
3. Vin Diesel & Paul Walker
4. Adam Sandler & Kevin James
5. Daniel Radcliffe & Rupert Grint

In [79]:
def unite (actor1, actor2):
    return len(data[data.cast.str.contains(actor1) & data.cast.str.contains(actor2)])
unite('Johnny Depp', 'Helena Bonham Carter')

6

In [80]:
unite('Hugh Jackman', 'Ian McKellen')

5

In [81]:
unite('Vin Diesel',  'Paul Walker')

5

In [82]:
unite('Adam Sandler', 'Kevin James')

5

In [83]:
unite('Daniel Radcliffe', 'Rupert Grint')

8

In [84]:
answer_ls.append(5)

# 36. У какого из режиссеров самый высокий процент фильмов со сборами выше бюджета?
Варианты ответа:
1. Quentin Tarantino
2. Steven Soderbergh
3. Robert Rodriguez
4. Christopher Nolan
5. Clint Eastwood

In [85]:
def director_profit(director):
    film = data[(data.director.str.contains(director))]
    profit_film = film[film.profit > 0]
    result = len(profit_film)*100/len(film)
    return result

directors = ['Quentin Tarantino', 'Steven Soderbergh', 'Robert Rodriguez', 'Christopher Nolan', 'Clint Eastwood']
for director in directors:
    print(director, '-', director_profit(director))

Quentin Tarantino - 85.71428571428571
Steven Soderbergh - 76.92307692307692
Robert Rodriguez - 72.72727272727273
Christopher Nolan - 100.0
Clint Eastwood - 83.33333333333333


In [86]:
answer_ls.append(4)

# Submission

In [87]:
len(answer_ls)

36

In [88]:
pd.DataFrame({'Id':range(1,len(answer_ls)+1), 'Answer':answer_ls}, columns=['Id', 'Answer'])

Id  Answer
0    1       4
1    2       2
2    3       3
3    4       2
4    5       1
5    6       5
6    7       2
7    8       1
8    9       4
9   10       5
10  11       3
11  12       1
12  13       3
13  14       4
14  15       5
15  16       1
16  17       3
17  18       3
18  19       2
19  20       1
20  21       4
21  22       2
22  23       3
23  24       1
24  25       5
25  26       1
26  27       4
27  28       2
28  29       5
29  30       2
30  31       5
31  32       5
32  33       3
33  34       1
34  35       5
35  36       4